cpp/dolfinx/fem/DirichletBC.h

```diff
@@ -495,76 +508,71 @@ class DirichletBC
   void set(std::span<T> x, std::optional<std::span<const T>> x0,
            T alpha = 1) const
   {
-    std::int32_t x_size = x.size();
-    if (alpha == T(0)) // Optimisation for when alpha == 0
+    // set_fn is a lambda which gets evaluated for every index in [0,
+    // _dofs0.size()) and its result is assigned to x[_dofs0[i]].
+    auto apply = [&](auto set_fn)
```

Comment:
Could we not use auto to define `set_fn`, but explicitly declare the type as
`std::function<T(std::size_t>)` ?

Reply:
And on that subject should `std::size_t` -> `std::int32_t` as all local dofs are `std::int32_t`?

Reply:
If the lambda does not capture a lambda will be inlinable trivially, but a `std::function` not, so I would refrain from using `std::function` in general here.

Reply:
But we could add a `std::invocable` concept to be more specific about what we allow.